pennylane/gradients/jvp.py

```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
+    #     - ((),..,())                             k scalar parameters
+    #     - [(l_1,..,l_{L_1}),..,(l_1,..,l_{L_k})] k mixed scalar and tensor parameters
+    #   jac:
+    #     - ((),..,())                             scalar return type and k scalar params
+    #     - ((r_1,..,r_R),..,(r_1,..,r_R))         tensor return type and k scalar params
+    #     - ((l_1,..,l_{L_1}),..,(l_1,..,l_{L_1})) scalar return type and k mixed params
+    #     - ((r_1,..,r_R, l_1,..,l_{L_1}),..,(r_1,..,r_R, l_1,..,l_{L_k}))
+    #                                              tensor return type and k mixed params
+    #
+    # Note that the return type dimensions always remain the same, whereas the
+    # dimensions of the tensor parameters may differ.
+    #
+    # In this scenario, another case separation is used: If any of the parameters is a
+    # tensor (i.e. not a scalar), all tangent entries are reshaped into one-dimensional
+    # tensors with shapes (prod_l_i,) and then stacked to one one-dimensional tensor.
+    # If there are no tensor parameters, the tangent is just stacked and reshaped.
+    # The Jacobians are reshaped to have the dimensions (r_1, ... r_R, prod_l_i)
+    # and then are concatenated along their last (potentially mismatching) axis.
+    # This is followed by a tensordot contraction over the axes of size sum_i prod_l_i.
+
     if jac is None:
         return None
-
-    tangent = qml.math.stack(tangent)
-    jac = _convert(jac, tangent)
-
-    # Single param
-    if not isinstance(jac, tuple):
+    single_param = not isinstance(jac, tuple)
+    if (single_param and jac.shape == (0,)) or (not single_param and len(jac) == 0):
         # No trainable parameters
-        if jac.shape == (0,):
-            res = qml.math.zeros((1, 0))
-            return res
-
-        tangent = qml.math.reshape(tangent, (1,))
-
-        # No dimension e.g. expval
-        if jac.shape == ():
-            jac = qml.math.reshape(jac, (1,))
-
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
+        num_l = len(tangent.shape[1:])
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
+        new_shape = (jac.shape if num_l == 0 else jac.shape[:-num_l]) + (prod_l,)
+        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
+        jac = qml.math.reshape(jac, new_shape)
+        return qml.math.tensordot(jac, tangent, [[-1], [0]])
+
+    if isinstance(tangent, (tuple, list)) and any(t.ndim > 0 for t in tangent):
+        num_l = [t.ndim for t in tangent]
+        tangent = [qml.math.reshape(t, (-1,)) for t in tangent]
+        prod_l = [t.shape[0] for t in tangent]
+        tangent = qml.math.hstack(tangent)
```

Comment:
Will values in `tangent` ever have different shapes? Because we might get an error when stacking.

Reply:
Yes they do initially but two lines up we change them to all be 1D :)

Reply:
This makes sense because whatever the shapes, we always want to contract over _all_ dimensions of the tangent, so to us it might as well be a single 1d object.

Reply:
But will they always have the same size? If they don't, `tangent` will be a list of 1D arrays with different sizes...

Reply:
Yep, they might, but that's not a problem, because we do `hstack`, not `stack` :)